### PR TITLE
Course module settings page

### PR DESCRIPTION
**Pull request by allenwq**
Similar to #156, Depends on #294

### CODE REVIEW

**Comment by lowjoel** [comment]
@allenwq Is this still wanted?

**Comment by allenwq** [comment]
@lowjoel  Yes, I've updated here, depends on #294 now

**Comment by lowjoel** [comment]
@allenwq please rebase

**Comment by allenwq** [comment]
@lowjoel I've updated here, thanks !

File: app/controllers/course/settings_controller.rb [hunk 1]
```diff
@@ -13,8 +14,27 @@ def update #:nodoc:
...
+      render 'components'
+    end
+  end
+
   private
 
+  # Load our settings adapter to handle component settings
+  def load_settings
+    @all_components = current_component_host.components + current_component_host.disabled_components
+    @settings = Course::Settings::Effective.new(@course, current_component_host)
```
Comment (lowjoel) [inline app/controllers/course/settings_controller.rb]: `current_course`

File: app/models/course/settings.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+class Course::Settings
```
Comment (lowjoel) [inline app/models/course/settings.rb]: How does this inherit settings from the instance?

File: spec/features/course/components_settings_spec.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        else
+          expect(checkbox).not_to be_checked
+        end
+      end
+    end
+
+    scenario 'Course owner can enable a component' do
+      visit course_components_path(course)
+
+      page.check(sample_component_id)
```
Comment (lowjoel) [inline spec/features/course/components_settings_spec.rb]: you can drop the `page`

**Comment by lowjoel** [comment]
I think there's too much logic in the controller.

**Comment by allenwq** [comment]
@lowjoel why there's too much logic ?

**Comment by allenwq** [comment]
@lowjoel  This is ready for review

File: app/controllers/components/course/component_host.rb [hunk 4]
```diff
@@ -111,8 +111,9 @@ def disabled_components
     instance_components - components
   end
 
-  private
-
+  # Returns the enabled components in instance.
+  #
+  # @return [Array] array of enabled components in instance
   def instance_components #:nodoc:
```
Comment (lowjoel) [inline app/controllers/components/course/component_host.rb]: Didn't remove?

File: app/controllers/course/settings_controller.rb [hunk 5]
```diff
@@ -13,8 +14,27 @@ def update #:nodoc:
...
+      @settings.errors = current_course.errors
+      render 'components'
+    end
+  end
+
   private
 
+  # Load our settings adapter to handle component settings
+  def load_settings
+    @instance_components = current_component_host.instance_components
```
Comment (lowjoel) [inline app/controllers/course/settings_controller.rb]: So why is this necessary now?

Comment (allenwq) [inline app/controllers/course/settings_controller.rb]: It's for views to use, you mean I should directly get this in views ?

File: spec/controllers/course/settings_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        all_component_ids = all_components.map { |c| c.key.to_s }
+        all_component_ids.sample(1 + rand(all_component_ids.count))
+      end
+      let(:components_params) { { enabled_component_ids: ids_to_enable } }
+
+      subject do
+        post :update_components, course_settings_effective: components_params, course_id: course
+      end
+
+      context 'enable/disable components' do
```
Comment (lowjoel) [inline spec/controllers/course/settings_controller_spec.rb]: why is this here? The context is unnecessary.

File: spec/controllers/course/settings_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        let(:settings) { Course::Settings.new(course.reload) }
+
+        it 'enables the specified component and disables all other components' do
+          subject
+          expect(settings.enabled_component_ids).to eq(ids_to_enable)
+        end
+      end
+
+      context 'when course cannot save' do
```
Comment (lowjoel) [inline spec/controllers/course/settings_controller_spec.rb]: when the course cannot be saved

File: spec/features/course/components_settings_spec.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
+require 'rails_helper'
+
+RSpec.feature 'Course: Components', type: :feature do
+  let!(:instance) { create(:instance) }
```
Comment (lowjoel) [inline spec/features/course/components_settings_spec.rb]: No need to declare `type: :feature`

Comment (allenwq) [inline spec/features/course/components_settings_spec.rb]: I see, thanks !

File: spec/controllers/course/settings_controller_spec.rb (hunk 7 above)
Comment (allenwq) [inline spec/controllers/course/settings_controller_spec.rb]: OK !

File: app/controllers/course/settings_controller.rb (hunk 5 above)
Comment (lowjoel) [inline app/controllers/course/settings_controller.rb]: Yes.

**Comment by allenwq** [comment]
@lowjoel Updated

File: app/views/course/settings/components.html.slim [hunk 9]
```diff
@@ -0,0 +1,15 @@
+h2 = t('.header')
+= simple_form_for @settings, url: course_components_path, method: :patch do |f|
```
Comment (lowjoel) [inline app/views/course/settings/components.html.slim]: I think :patch should be deduced.

File: app/views/course/settings/components.html.slim [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+  table.table
+    thead
+      th = t('.name')
+      th = t('.enabled')
+    tbody
+      - instance_components = controller.current_component_host.instance_components
```
Comment (lowjoel) [inline app/views/course/settings/components.html.slim]: is current_component_host a helper method?

File: config/routes.rb [hunk 11]
```diff
@@ -90,6 +90,8 @@
 
       get 'settings' => 'settings#index', as: :settings
       patch 'settings' => 'settings#update', as: :update_settings
```
Comment (lowjoel) [inline config/routes.rb]: This is your old PR, but the :as should not be needed.

File: app/views/course/settings/components.html.slim (hunk 10 above)
Comment (allenwq) [inline app/views/course/settings/components.html.slim]: No, you said don't need to add a `helper_method ` to  `current_component_host ` before :)

File: config/routes.rb (hunk 11 above)
Comment (allenwq) [inline config/routes.rb]: I already realised, will fix this in another PR

File: app/controllers/course/settings_controller.rb [hunk 12]
```diff
@@ -13,8 +14,26 @@ def update #:nodoc:
     end
   end
 
+  def components #:nodoc:
+  end
+
+  def update_components #:nodoc:
+    @settings.update(params.require(:settings_effective))
```
Comment (lowjoel) [inline app/controllers/course/settings_controller.rb]: should call this `settings_effective_params`

Comment (lowjoel) [inline app/controllers/course/settings_controller.rb]: actually, update calls save...

Comment (lowjoel) [inline app/controllers/course/settings_controller.rb]: wait a minute...

Comment (lowjoel) [inline app/controllers/course/settings_controller.rb]: sorry, mistook.

Comment (allenwq) [inline app/controllers/course/settings_controller.rb]: The param name is same as instance yah ?
https://github.com/Coursemology/coursemology2/blob/master/app/controllers/admin/admin_controller.rb#L12

File: spec/features/course/components_settings_spec.rb [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  with_tenant(:instance) do
+    let(:user) { create(:user) }
+    let(:course) { create(:course, creator: user) }
+    let(:components)  { Course::ComponentHost.components }
+    let(:sample_component_id) do
+      "settings_effective_enabled_component_ids_#{components.sample.key}"
+    end
+    before { login_as(user, scope: :user) }
+
+    scenario 'Course owner can view the list of enabled/disabled components' do
```
Comment (lowjoel) [inline spec/features/course/components_settings_spec.rb]: Maybe we should use context for 'Course Owner' or 'Administrator' or 'User'

**Comment by allenwq** [comment]
@lowjoel Updated !
